pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java

```diff
@@ -0,0 +1,468 @@
+    _forceEnableDictEncoding = true;
+  }
+
+  /**
+   * Forces the use of raw encoding, overriding any automatic encoding decisions.
+   *
+   * <p><b>Note:</b> This method is exclusive to {@code forceClpDictionaryEncoding}; enabling raw encoding will
+   * disable clp dictionary encoding. Only one of these methods can be active at a time.</p>
+   */
+  public void forceRawEncoding() {
+    _isClpEncoded = false;
+    _bytesRawFwdIndexDocIdStartOffset = 0;
+  }
+
+  public String getColumnName() {
+    return _columnName;
+  }
+
+  @Override
+  public String getString(int docId) {
+    return new String(getRawBytes(docId), StandardCharsets.UTF_8);
+  }
+
+  public byte[] getRawBytes(int docId) {
+    if (docId < 0) {
+      throw new IllegalArgumentException("Invalid docId: " + docId);
+    }
+    if (docId < _bytesRawFwdIndexDocIdStartOffset) {
+      // Decode from clp column
+      byte[] logtype = _logtypeDict.get(_logtypeId.getInt(docId));
+
+      int dictVarIdBeginOffset = (0 == docId) ? 0 : _dictVarOffset.getInt(docId - 1);
+      int dictVarIdEndOffset = _dictVarOffset.getInt(docId);
+      byte[][] dictVars = new byte[dictVarIdEndOffset - dictVarIdBeginOffset][];
+      for (int i = 0; i < dictVars.length; i++) {
+        dictVars[i] = _dictVarDict.get(_dictVarId.getInt(dictVarIdBeginOffset + i));
+      }
+      FlattenedByteArray flattenedDictVars = FlattenedByteArrayFactory.fromByteArrays(dictVars);
+
+      int encodedVarIdBeginOffset = (0 == docId) ? 0 : _encodedVarOffset.getInt(docId - 1);
+      int encodedVarIdEndOffset = _encodedVarOffset.getInt(docId);
+      long[] encodedVars = new long[encodedVarIdEndOffset - encodedVarIdBeginOffset];
+      for (int i = 0; i < encodedVars.length; i++) {
+        encodedVars[i] = _encodedVar.getLong(encodedVarIdBeginOffset + i);
+      }
+
+      try {
+        return _clpMessageDecoder.decodeMessageAsBytes(logtype, flattenedDictVars, encodedVars);
+      } catch (IOException e) {
+        throw new IllegalArgumentException(
+            "Failed to encode message: " + new String(logtype, StandardCharsets.ISO_8859_1), e);
+      }
+    } else {
+      return _rawBytes.getBytes(docId - _bytesRawFwdIndexDocIdStartOffset);
+    }
+  }
+
+  public FixedByteSVMutableForwardIndex getLogtypeId() {
+    return _logtypeId;
+  }
+
+  public BytesOffHeapMutableDictionary getLogtypeDict() {
+    return _logtypeDict;
+  }
+
+  public FixedByteSVMutableForwardIndex getDictVarOffset() {
+    return _dictVarOffset;
+  }
+
+  public FixedByteSVMutableForwardIndex getDictVarId() {
+    return _dictVarId;
+  }
+
+  public BytesOffHeapMutableDictionary getDictVarDict() {
+    return _dictVarDict;
+  }
+
+  public FixedByteSVMutableForwardIndex getEncodedVarOffset() {
+    return _encodedVarOffset;
+  }
+
+  public FixedByteSVMutableForwardIndex getEncodedVar() {
+    return _encodedVar;
+  }
+
+  /**
+   * Returns whether the mutable forward index is currently using dictionary encoding.
```

Comment:
> whether the mutable forward index is currently using dictionary encoding
whether the mutable forward index is currently using CLP forward indexes with dictionary encoding or rawFwdIndex.